### ISSUE TITLE
Support for Criteria equivalent based queries like JpaSpecificationExecutor

### ISSUE THREAD

**Issue by rcpacheco** [spring-projects/spring-data-r2dbc#246]
Support for Criteria equivalent funtionality to do queries like JpaSpecificationExecutor.
Being a user of spring data jpa, we use a lot the Specification based queries to implement
paging, sort, search and filter funtionality in our apps. Having a similar capability would make the migration to r2dbc easier.

**Comment by LaoTsing** [comment]
Yes, supporting custom Criteria make functional safe and clean

**Comment by guillaumelamirand** [comment]
Hi,

I agree with this. How could use queries like JpaSpecificationExecutor with spring data r2dbc ? I tried to mix data jpa and data r2dbc only for pagination queries but the application won't start. Do you have an ETA for that feature ? 

Thanks

**Comment by mp911de** [comment]
I'm not sure we want to introduce such a feature in the first place. `R2dbcEntityTemplate` exposes exactly what you're looking for. What's wrong with using `R2dbcEntityTemplate.select(Query, Class)`?

**Comment by guillaumelamirand** [comment]
Yes of course but from my point of view it is also the purpose of a Repository to offer those kind of features. First of all because it is easier to use and also it is how we already use it with standard JPA implementation.

Why don't you want to add such feature? How whould we with r2dbc repository manage pagination and search ?

**Comment by mp911de** [comment]
You can use `Pageable` and `Sort` with query derivation. Alternatively, you can use `@Query` to specify your query.

Again, what's wrong with using `R2dbcEntityTemplate.select(Query, Class)`?

**Comment by guillaumelamirand** [comment]
In fact we are not using Query but Predicate coming from a built Specification, not sure R2dbcEntityTemplate.select(Query, Class) can solve this.

**Comment by mp911de** [comment]
Note that Spring Data R2DBC doesn't use JPA API hence there's no support for `javax.persistence.criteria.Predicate`.

**Comment by guillaumelamirand** [comment]
Ok, that confirm what I suspected. Does it mean I cannot use criteria api with R2dbc?

**Comment by mp911de** [comment]
There's `org.springframework.data.relational.core.query.Criteria` which is obviously different from the JPA `Criteria` type. However, they work similar and I'd argue that `org.springframework.data.relational.core.query.Criteria` is simpler to use.

**Comment by guillaumelamirand** [comment]
We are building a complex specification from a search param given by user to request what he wants. He gives us sth like "/topics?search=type:toto AND channel.user?" and we transform that into a Speficiation used to request the db.

I will have a look if `org.springframework.data.relational.core.query.Criteria` can be used the same way.
Thanks any way for helping.

**Comment by rcpacheco** [comment]
We have the same situation, using specs to filter data from grids makes our work more Easy.

**Comment by guillaumelamirand** [comment]
@mp911de Sorry for the delay. I did take the time today to look into R2dbcEntityTemplate, Query and Criteria but from what I understand query doesn't manage entities relationship (join, left join, etc) so how could I use it with a complex query joining multiple table ?

**Comment by guillaumelamirand** [comment]
@mp911de ok . Is it something plan for future release?

**Comment by encircled** [comment]
Btw Hibernate introduced completable future reactive API via Vert.x driver (and I've implemented a criteria API over it https://github.com/encircled/Joiner with support of reactor mono/flux)

**Comment by hantsy** [comment]
Micronaut Data R2dbc has already supported JPA annotations(on entity and associations) and such an JPA Specification variant(for complex queries) in the R2dbc(and Kotlin Coroutines) directly. I remember I have filed similar issues to request these features in Spring Data Jdbc/R2dbc on Spring Data Relational, but rejected.

**Comment by mp911de** [comment]
JPA annotations have a very different semantic associated and a lot of these (e.g. `@GeneratedValue`, `@OneToOne`, and others) are not supported by Spring Data R2DBC. Supporting JPA annotations would send out a bold signal to users and we immediately need to manage expectations to reduce the supported set of features so that it does not make any sense to support JPA annotations in the first place.
